tensorboard/plugins/debugger/session_debug_test.py

```diff
@@ -0,0 +1,339 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests end-to-end debugger data server behavior by starting TensorBoard.
+
+This test launches an instance of TensorBoard as a subprocess. In turn,
+TensorBoard (specifically its debugger plugin) starts a debugger data server.
+The test then calls Session.run() using RunOptions pointing to the grpc:// debug
+URL of the debugger data server. It then checks the correctness of the Event
+proto file created by the debugger data server.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import functools
+import glob
+import os
+import shutil
+import tempfile
+import threading
+import time
+
+import numpy as np
+import portpicker
+import tensorflow as tf
+
+from tensorflow.python import debug as tf_debug
+
+from tensorboard.plugins.debugger import constants
+from tensorboard.plugins.debugger import debugger_server_lib
+
+
+class SessionDebugTestBase(tf.test.TestCase):
+
+  def setUp(self):
+    self._debugger_data_server_grpc_port = portpicker.pick_unused_port()
+    self._debug_url = (
+        "grpc://localhost:%d" % self._debugger_data_server_grpc_port)
+    self._logdir = tempfile.mkdtemp(prefix="tensorboard_dds_")
+
+    self._debug_data_server = debugger_server_lib.DebuggerDataServer(
+        self._debugger_data_server_grpc_port, self._logdir, always_flush=True)
+    self._server_thread = threading.Thread(
+        target=self._debug_data_server.start_the_debugger_data_receiving_server)
+    self._server_thread.start()
+
+    self.assertTrue(self._poll_server_till_success(50, 0.2))
+
+  def tearDown(self):
+    self._debug_data_server.stop_server()
+    self._server_thread.join()
+
+    if os.path.isdir(self._logdir):
+      shutil.rmtree(self._logdir)
+
+    tf.reset_default_graph()
+
+  def _poll_server_till_success(self, max_tries, poll_interval_seconds):
+    poll_success = False
+    for _ in range(max_tries):
+      try:
+        with tf.Session() as sess:
+          a_init_val = np.array([42.0])
+          a_init = tf.constant(a_init_val, shape=[1], name="a_init")
+          a = tf.Variable(a_init, name="a")
+
+          run_options = tf.RunOptions(output_partition_graphs=True)
+          tf_debug.watch_graph(run_options,
+                               sess.graph,
+                               debug_ops=["DebugNumericSummary"],
+                               debug_urls=[self._debug_url])
+
+          sess.run(a.initializer, options=run_options)
+          poll_success = True
```

Comment:
Might as well just `return True` here and `return False` outside the `for` loop. Less state.